network/network_linux.go

```diff
@@ -345,23 +355,34 @@ func applyDnsConfig(extIf *externalInterface, ifName string) error {
 				continue
 			}
 
-			buf := fmt.Sprintf("--set-dns=%s", server)
-			setDnsList = setDnsList + " " + buf
+			setDnsList = setDnsList + " " + server
 		}
 
 		if setDnsList != "" {
-			cmd := fmt.Sprintf("systemd-resolve --interface=%s%s", ifName, setDnsList)
-			_, err = p.ExecuteCommand(cmd)
+			// example command on Ubuntu22: resolvectl dns azure0 168.63.129.16
+			resolveCtlCmd := fmt.Sprintf("resolvectl dns %s%s", ifName, setDnsList)
+			// example command on Ubuntu18: systemd-resolve --interface=azure0 --set-dns=168.63.129.16
+			systemdResolveCmd := fmt.Sprintf("systemd-resolve --interface=%s%s", ifName, "--set-dns="+strings.TrimSpace(setDnsList))
+
+			_, err = p.ExecuteCommand(resolveCtlCmd)
+			log.Errorf("resolvectl cmd failed with:%v",err) //nolint
 			if err != nil {
+				_, err = p.ExecuteCommand(systemdResolveCmd)
 				return err
 			}
 		}
 
 		if extIf.DNSInfo.Suffix != "" {
-			cmd := fmt.Sprintf("systemd-resolve --interface=%s --set-domain=%s", ifName, extIf.DNSInfo.Suffix)
-			_, err = p.ExecuteCommand(cmd)
+			// example command on Ubuntu22: resolvectl domain azure0 dlw5dhyl2njevcuzgmfubi2oid.bx.internal.cloudapp.net
+			resolveCtlCmd := fmt.Sprintf("resolvectl domain %s %s", ifName, extIf.DNSInfo.Suffix)
+			// example command on Ubuntu18: systemd-resolve --interface=azure0 --set-domain=dlw5dhyl2njevcuzgmfubi2oid.bx.internal.cloudapp.net
+			systemdResolveCmd := fmt.Sprintf("systemd-resolve --interface=%s --set-domain=%s", ifName, extIf.DNSInfo.Suffix)
+
+			if _, err = p.ExecuteCommand(resolveCtlCmd); err != nil {
+				log.Errorf("resolvectl cmd failed with:%v",err) //nolint
+				_, err = p.ExecuteCommand(systemdResolveCmd)
+			}
```

Comment:
i doubt this is safe in scenarios where both systemd-resolve and resolvectl are available, and the first one has a legitimate error